rewrote the escape function to escape all markdown characters

yareeba opened the pull request:
in order to fix issue #233

Commit:
rewrote the escape function to escape all markdown characters

domchristie commented on src/turndown.js:
```diff
@@ -6,6 +6,24 @@ import Node from './node'
 var reduce = Array.prototype.reduce
 var leadingNewLinesRegExp = /^\n*/
 var trailingNewLinesRegExp = /\n*$/
+const markdownReplacements = [
```
Perhaps `var escapes = [`

Paul-Ladyman replied:
done

domchristie commented on src/turndown.js:
```diff
@@ -6,6 +6,24 @@ import Node from './node'
+  [/-/g, '\\-'],
+  [/\+/g, '\\+'],
+  [/=/g, '\\='],
+  [/#/g, '\\#'],
```
`#` is only used for atx-style headings, so we can be more specific and reduce the number of escapes. Something like:

```js
[/^(\W* {0,3})(#{1,6} )/gm, '$1\\$2']
```

Paul-Ladyman replied:
Simplified this to

```javascript
[/^(#{1,6})/g, '\\$1']
```

produces an escaped headline like this:

`### headline` = `\### headline`

domchristie replied:
Yes, a single backslash if sufficient: https://spec.commonmark.org/dingus/?text=%5C%23%23%23%20Not%20a%20heading%0A

domchristie commented on src/turndown.js:
```diff
@@ -6,6 +6,24 @@ import Node from './node'
+  [/=/g, '\\='],
+  [/#/g, '\\#'],
+  [/`/g, '\\`'],
+  [/~/g, '\\~'],
```
`~` can only be used for code blocks, so to reduce the number of escapes, something like:

```js
[/^(\W* {0,3})~~~/gm, '$1\\~~~']
```

Paul-Ladyman replied:
It can also be used for a strikethrough effect:

`~~strikethrough~~` = ~~strikethrough~~

In this case shall we stick with the aggressive escaping for ~?

domchristie commented on src/turndown.js:
```diff
@@ -6,6 +6,24 @@ import Node from './node'
+  [/#/g, '\\#'],
+  [/`/g, '\\`'],
+  [/~/g, '\\~'],
+  [/\|/g, '\\|'],
```
As pipes (`|`) are used for tables, this should probably be part of turndown-plugin-gfm, and so can be removed

domchristie commented on src/turndown.js:
```diff
@@ -6,6 +6,24 @@ import Node from './node'
+  [/~/g, '\\~'],
+  [/\|/g, '\\|'],
+  [/\(/g, '\\('],
+  [/\)/g, '\\)'],
```
I don't think `(` or `)` need to be escaped if `[` and `]` are escaped, so we can remove these two lines

Paul-Ladyman replied:
done

domchristie commented on src/turndown.js:
```diff
@@ -6,6 +6,24 @@ import Node from './node'
+  [/\*/g, '\\*'],
+  [/-/g, '\\-'],
+  [/\+/g, '\\+'],
+  [/=/g, '\\='],
```
`=` is only used for setext headings so we can be more specific and reduce the number of escapes by doing something like:

```js
[/^(\W* {0,3})(=+)/gm, '$1\\$2']
```

Paul-Ladyman replied:
Again simplified to:

```javascript
[/^(=+)/g, '\\$1']
```

Giving:

`===` = `\===`

domchristie replied:
Seems to do the trick: https://spec.commonmark.org/dingus/?text=Not%20a%20heading%0A%5C%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%0A

domchristie commented on src/turndown.js:
```diff
@@ -6,6 +6,24 @@ import Node from './node'
+  [/\]/g, '\\]'],
+  [/>/g, '\\>'],
+  [/_/g, '&lowbar;'],
+  [/(\d+)\./g, '$1\\.']
```
Digits followed by a `.` can only be markdown if they are at the start of a line, so perhaps something like:

```js
[/^(\W* {0,3})(\d+)\. /gm, '$1$2\\. ']
```

Paul-Ladyman replied:
done

domchristie commented on src/turndown.js:
```diff
@@ -6,6 +6,24 @@ import Node from './node'
+  [/\)/g, '\\)'],
+  [/\[/g, '\\['],
+  [/\]/g, '\\]'],
+  [/>/g, '\\>'],
```
To reduce the number of escapes, perhaps:

```js
[/^(\W* {0,3})> /gm, '$1\\> ']
```

Paul-Ladyman replied:
done

domchristie commented on src/turndown.js:
```diff
@@ -6,6 +6,24 @@ import Node from './node'
+const markdownReplacements = [
+  [/\\/g, '\\\\'],
+  [/\*/g, '\\*'],
+  [/-/g, '\\-'],
```
This is a little tricky because `-` can be used both for bullets and hrs. We might need to allow function replacements. e.g.:

```js
[
  /^\W* {0,3}(- *)+/gm,
  function (match, character) {
    return match.split(character).join('\\' + character)
  }
]
```

Paul-Ladyman replied:
Do you think as a first pass it would be reasonable to escape - aggressively and work on more sophisticated escaping as part of a future task?

domchristie commented on src/turndown.js:
```diff
@@ -126,47 +144,9 @@ TurndownService.prototype = {
   escape: function (string) {
-    return (
+    return markdownReplacements.reduce(
+      (search, replacement) => search.replace(replacement[0], replacement[1]),
```
Just to clarify this, what do you reckon about:

```js
escapes.reduce(function (accumulator, escape) {
  return accumulator.replace(escape[0], escape[1])
}, string)
```

?

Paul-Ladyman replied:
I definitely prefer the variable names. Is there a particular reason not to use the Lambda syntax? Are you worried about Node version compatibility or is it just a question of style? I notice similarly you suggested we prefer using `var` to `const`. Same reasoning?

domchristie replied:
> Is there a particular reason not to use the Lambda syntax? Are you worried about Node version compatibility or is it just a question of style? I notice similarly you suggested we prefer using var to const. Same reasoning?

The library also targets the browser and aims for broad compatibility without having to depend on a transpiler and so it is written in ES5 (ES6 modules for building distributables).

domchristie commented on test/index.html:
```diff
@@ -665,38 +665,38 @@ <h2>This is a header.</h2>
 <div class="case" data-name="escaping headings with #">
   <div class="input">### This is not a heading</div>
-  <pre class="expected">\### This is not a heading</pre>
+  <pre class="expected">\#\#\# This is not a heading</pre>
```
This can be reverted

domchristie commented on src/turndown.js:
```diff
@@ -6,6 +6,24 @@ import Node from './node'
+  [/\[/g, '\\['],
+  [/\]/g, '\\]'],
+  [/>/g, '\\>'],
+  [/_/g, '&lowbar;'],
```
Is there a reason why this is `&lowbar;` rather than `\\_`?

Paul-Ladyman replied:
Yeah this is not great on our part. As part of our stack we use `turndown` to turn HTML to markdown and then later [`marked`](https://github.com/markedjs/marked) to turn the markdown back to HTML.

We were having a weird issue where escaping `_` as `\_` was not then being parsed correctly when going back to HTML. Probably a double escaping issue as the `\` was being rendered to the screen (weirdly we don't have this issue any other time we escape using `\`). Our solution for now in our code is to use `&lowbar;` instead.

Obviously we shouldn't let our specific problem affect `turndown` but we left it there for now whilst we were copying our code here.

So there's probably three options:
* leave it how it is (using `&lowbar;`), at least for now.
* Use HTML character codes everywhere instead of `\` for escaping. A little outlandish but maybe makes sense since the output is HTML anyway.
* Use `\_` in `turndown` instead and we find another approach to our issue. This probably involves us sticking with overloading `escape` for now despite this PR so from that point of view it seems a shame.

Sorry for the long reply. Any thoughts either way?

domchristie commented on test/index.html:
```diff
@@ -729,9 +729,9 @@ <h2>This is a header.</h2>
+<div class="case" data-name="escaping numbers followed by a full stop in a sentence">
   <div class="input">George Orwell wrote 1984.</div>
-  <pre class="expected">George Orwell wrote 1984.</pre>
+  <pre class="expected">George Orwell wrote 1984\.</pre>
```
This test case should be reverted

domchristie commented on test/index.html:
```diff
@@ -749,29 +749,59 @@ <h2>This is a header.</h2>
+<div class="case" data-name="escaping =">
+  <div class="input">A sentence containing =</div>
+  <pre class="expected">A sentence containing \=</pre>
 </div>
```
This test case can probably be removed

domchristie commented on test/index.html:
```diff
@@ -749,29 +749,59 @@ <h2>This is a header.</h2>
+<div class="case" data-name="escaping > within text">
   <div class="input">42 > 1</div>
-  <pre class="expected">42 > 1</pre>
+  <pre class="expected">42 \> 1</pre>
```
This test case can be reverted

domchristie commented on test/index.html:
```diff
@@ -749,29 +749,59 @@ <h2>This is a header.</h2>
+<div class="case" data-name="escaping parentheses">
+  <div class="input">(A sentence containing)</div>
+  <pre class="expected">\(A sentence containing\)</pre>
+</div>
```
This test case can be removed

domchristie commented on test/index.html:
```diff
@@ -749,29 +749,59 @@ <h2>This is a header.</h2>
+<div class="case" data-name="escaping |">
+  <div class="input">A sentence containing |</div>
+  <pre class="expected">A sentence containing \|</pre>
+</div>
```
This test case can be removed

domchristie commented on test/index.html:
```diff
@@ -749,29 +749,59 @@ <h2>This is a header.</h2>
+<div class="case" data-name="escaping ~">
+  <div class="input">A sentence containing ~</div>
+  <pre class="expected">A sentence containing \~</pre>
+</div>
```
This should altered to only test a case where a `~` could be converted to markdown

domchristie reviewed (changes requested):
Thanks for your contribution! 💫 I love the use of `reduce`!

I have made quite a few comments, so I hope this doesn't come across as ungrateful…

Overall, I like this approach. As I mentioned before, I think the library should escape aggressively to ensure that valid markdown is outputted. However, I don't think it is necessary to escape all markdown characters in any context (sorry, I should've mention this before). Escaping `*`, `_`, and <code>\`</code> in all contexts makes sense, but for the other characters, I think we can be a little less aggressive so that the output isn't overloaded with `\`.

As I said, I hope the comments aren't too overwhelming. Let me know what you think, and thanks again!

Commit:
Changes after code review. Escape certain characters only when used in valid markdown.

Paul-Ladyman commented on src/turndown.js:
```diff
@@ -6,6 +6,24 @@ import Node from './node'
+  [/\\/g, '\\\\'],
+  [/\*/g, '\\*'],
+  [/-/g, '\\-'],
+  [/\+/g, '\\+'],
```
done

Commit:
Changes after code review

- Don't escape | here as its not commonmark
- Use regular escaping for _ rather than HTML entities

Commit:
Changes after code review. Only escape ~ if used as code block.
